libvast/src/logger.cpp

```diff
@@ -15,9 +15,250 @@
+create_log_context(const vast::invocation& cmd_invocation,
+                   const caf::settings& cfg_file) {
+  if (!vast::detail::setup_spdlog(cmd_invocation, cfg_file))
+    return {};
+
+  return {caf::detail::make_scope_guard(
+    std::addressof(vast::detail::shutdown_spdlog))};
+}
+
+namespace {
+
+constexpr bool is_vast_loglevel(const int value) {
+  switch (value) {
+    case VAST_LOG_LEVEL_QUIET:
+    case VAST_LOG_LEVEL_ERROR:
+    case VAST_LOG_LEVEL_WARNING:
+    case VAST_LOG_LEVEL_INFO:
+    case VAST_LOG_LEVEL_VERBOSE:
+    case VAST_LOG_LEVEL_DEBUG:
+    case VAST_LOG_LEVEL_TRACE:
+      return true;
+  }
+  return false;
+}
+
+
+/// Converts a vast log level to spdlog level
+constexpr spdlog::level::level_enum vast_loglevel_to_spd(const int value) {
+  assert(is_vast_loglevel(value));
+
+  spdlog::level::level_enum level = spdlog::level::off;
+  switch (value) {
+    case VAST_LOG_LEVEL_QUIET:
+      break;
+    case VAST_LOG_LEVEL_ERROR:
+      level = spdlog::level::err;
+      break;
+    case VAST_LOG_LEVEL_WARNING:
+      level = spdlog::level::warn;
+      break;
+    case VAST_LOG_LEVEL_INFO:
+      level = spdlog::level::info;
+      break;
+    case VAST_LOG_LEVEL_VERBOSE:
+      level = spdlog::level::debug;
+      break;
+    case VAST_LOG_LEVEL_DEBUG:
+      level = spdlog::level::debug;
+      break;
+    case VAST_LOG_LEVEL_TRACE:
+      level = spdlog::level::trace;
+      break;
+  }
+  return level;
+}
+
+// There are log calls done before the log was created ,
+// so something to ignore is needed
+inline std::shared_ptr<spdlog::async_logger> dev_null_logger() {
+  auto null_logger
+    = spdlog::async_factory::template create<spdlog::sinks::null_sink_mt>(
+      "/dev/null");
+  null_logger->set_level(spdlog::level::off);
+  return null_logger;
+}
+
+std::shared_ptr<spdlog::async_logger> vast_logger = dev_null_logger();
+
+} // namespace
+
+namespace detail {
+
+bool setup_spdlog(const vast::invocation& cmd_invocation,
+                  const caf::settings& cfg_file) {
+  // already running .....
+  if (vast_logger && vast_logger->name() != "/dev/null") {
+    VAST_ERROR_ANON("Log already up");
+    return false;
+  }
+
+  bool is_server = cmd_invocation.name() == "start"
+                   || caf::get_or(cmd_invocation.options, "vast.node", false);
+
+  const auto& cfg_cmd = cmd_invocation.options;
+
+  auto console_verbosity = vast::defaults::logger::console_verbosity ;
+  auto file_verbosity = vast::defaults::logger::file_verbosity ;
+
+  // The log level setup, just top down.
+  // Command line overwrites config file values
+  auto verbosity = caf::get_if<caf::atom_value>(&cfg_cmd, "vast.verbosity");
+  if (verbosity) {
+    if (loglevel_to_int(*verbosity, -1) < 0) {
+      std::cerr<< "Illegal vast.verbosity " << to_string(*verbosity) << "\n" ;
+      return false;
+    }
+    file_verbosity = *verbosity ;
+    console_verbosity = *verbosity;
+  } else {
```

Comment:
It follows the latest specification I got from here, this was no comneted different from you: 
https://matrix.to/#/!bVBdWibynPCEuKUlIT:tenzir.modular.im/$fx3LLIn4JyjZZjsM0C_whXL1UczPbQmPr_--Y2kzhrA?via=tenzir.modular.im&via=matrix.org

```
Harald
I don't want to introduce that, but we had it as topic,
I would like to have just vast.verbosity, vast.file-verbosity, and vast.console-verbosity
vast.verbosiy comes from the command line and if provided overwrites the other both
vast.file-verbosity, and vast.console-verbosity come from the config file, if they are missing the defaults strike
I think this is simple and straight forward, can we agree on that?

dominiklohmann
Yes, and it matches what we currently have in place.

```

Reply:
Will test if in the old implementation command line overwrites also the file logging, if not, implementation will be adopted

Reply:
Perfect. Fwiw, I think @dominiklohmann was looking more into the structure of your proposal (ie. `vast.verbosity` instead of `vast.logger.verbosity`) and not the semantics of what the options do, and I also wasn't reading your comment deeply enough to catch this immediately. Sorry for that :/

Reply:
Cross checked wit previous implementation, the `-vvv` did not change file verbosity.
So I took that behaviour over.